Add logging to component

michalbachowski opened the pull request:
flask_cors is really great component, but it has virtually no logging, which makes it hard to debug.
I've forked project and added some logging messages here and there. Also created `_stringify_regexp()` helper function that returns pattern representation for given regexp / string. I think that more messages will be needed in the future, but this PR might be a good start.

There is one more thing to think about - whether to add or not default handler (NullHandler), as discussed in [A Brief Digression About Logging](https://lukasa.co.uk/2014/05/A_Brief_Digression_About_Logging/) blog post. Current state will trigger "warning" under Python 2.

Best regards

michalbachowski commented:
I see travis errors under Python 2.6. In case of accepting this PR I will fix it by dropping `format()` in favour of old percent formatting.

CoryDolphin commented:
Michael, this is wonderful! I have wanted to add logging for quite some
time but haven't had a chance to really dig into it.

I will plan to spend some time digging in and seeing where else I think
logging would make sense. This is a great contribution, thank you very much
for taking the time!
On Mon, Jan 19, 2015 at 6:35 AM Michał Bachowski <notifications@github.com>
wrote:

> I see travis errors under Python 2.6. In case of accepting this PR I will
> fix it by dropping format() in favour of old percent formatting.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CoryDolphin/flask-cors/pull/99#issuecomment-70480315>.
>

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -150,6 +153,7 @@ def wrapped_function(*args, **kwargs):
             options.update(_get_app_kwarg_dict())
             options.update(_options)
             _serialize_options(options)
+            logger.debug('Final CORS options: {}'.format(str(options)))
```
My vote is to move this logging statement to Cors. init_app. This will get logged on every invocation of the route otherwise. And for the decorator, there is not actually any sense of finality, since this computed at request invocation.

michalbachowski replied:
Correct me if I'm wrong, but decorator will not call init_app() on it's own, so it is possible to initiate CORS on specific route without instantinating CORS() class. In such case logging information will be lost.

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -305,9 +332,11 @@ def _get_cors_origin(options, request_origin):
     # If the Origin header is not present terminate this set of steps.
     # The request is outside the scope of this specification.-- W3Spec
     if request_origin:
+        logger.debug("'Origin' header was send with value '{}'".format(str(request_origin)))
```
Sp on 'send' -> 'sent'

michalbachowski replied:
+1

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -342,7 +377,9 @@ def _get_cors_headers(options, request_headers, request_method,
     origin_to_set = _get_cors_origin(options, request_headers.get('Origin'))
 
     if origin_to_set is None:  # CORS is not enabled for this route
+        logger.debug("Could not obtain value of origin header for response, skipping")
```
I think this is overly verbose. The log on https://github.com/CoryDolphin/flask-cors/pull/99/files#diff-9510ebf5b86a1db79a60dbe1e9fe2d71R370 should be perfect for this

michalbachowski replied:
+1

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -381,12 +418,14 @@ def _set_cors_headers(resp, options):
                                        request.headers,
                                        request.method,
                                        resp.headers)
+    logger.debug('CORS headers to be set {}'.format(str(headers_to_set)))
```
I think it would be great to bump this up to an INFO level, and perhaps change the tensing to "Settings CORS headers: ...". Thoughts?

michalbachowski replied:
+1

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -330,9 +362,12 @@ def _get_cors_origin(options, request_origin):
     # This will only be exceuted it there is no Origin in the request, in which
     # case, why bother with CORS?
     elif options.get('always_send') and options.get('origins_str'):
+        logger.debug("Althought 'Origin' header was not send, we are going " + \
```
Proposed change:
```"'always_send' is True. Settings headers even though 'Origin' header is missing".```

michalbachowski replied:
+1

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -330,9 +362,12 @@ def _get_cors_origin(options, request_origin):
         return options.get('origins_str')
     # Terminate these steps, return the original request untouched.
     else:
+        logger.debug("'Origin' header was not send, which means CORS was not requested, skipping")
```
Proposed wording change:

```"No 'Origin' header in request, skipping"```

michalbachowski replied:
+1

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -238,17 +244,25 @@ def cors_after_request(resp):
                 The actual after-request handler, retains references to the
                 the options, and definitions of resources through a closure.
             '''
+            logger.debug('Checking CORS for request')
```
I think this is overly verbose. I think as a goal, we should only log stuff when we know we are doing work, and only when it will be somewhat unique to the request.

michalbachowski replied:
+1

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -238,17 +244,25 @@ def cors_after_request(resp):
+            logger.debug('Checking CORS for request')
             # If CORS headers are set in a view decorator, pass
             if resp.headers.get(ACL_ORIGIN):
+                logger.debug('Header {} have been set already, skipping'.format(ACL_ORIGIN))
```
I propose using the same comment as below:
```logger.debug('CORS have been already evaluated, skipping')```

michalbachowski replied:
+1

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -238,17 +244,25 @@ def cors_after_request(resp):
+                logger.debug('Header {} have been set already, skipping'.format(ACL_ORIGIN))
                 return resp
 
+            logger.debug('Checking available CORS rules')
```
I also suggest removing this line, as it will be written out on every request :-(

michalbachowski replied:
+1

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -238,17 +244,25 @@ def cors_after_request(resp):
                     _options = options.copy()
                     _options.update(res_options)
                     _serialize_options(_options)
+                    logger.debug('Final CORS options: {}'.format(str(_options)))
```
Can we combine these two logging statements?

E.G:

```Python
logger.debug("Cors resource '%s' matches. Using options: %s", _stringify_regexp(res_regex), res_options)
```

michalbachowski replied:
+1. Just removed duplicated "matches".

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -238,17 +244,25 @@ def cors_after_request(resp):
                     break
+            else:
+                logger.debug('No CORS rule matches')
+            logger.debug('Response headers: {}'.format(str(resp.headers)))
```
I think this log statement is duplicated when we actually set the headers. I vote to remove this one, and keep the other

michalbachowski replied:
+1

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -157,6 +161,7 @@ def wrapped_function(*args, **kwargs):
                 resp = make_response(f(*args, **kwargs))
 
             _set_cors_headers(resp, options)
+            logger.debug('Response headers: {}'.format(str(resp.headers)))
```
I think we should remove this log, as it will be logging a lot of information not specific to CORS, and our specific information is already logged in the set_cors_headers call.

michalbachowski replied:
+1

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -229,6 +234,7 @@ def init_app(self, app, **kwargs):
         options.update(_get_app_kwarg_dict(app))
         options.update(self._options)
         options.update(kwargs)
+        logger.debug('setting CORS for app {}'.format(app.name))
```
I propose moving this logging statement below the call to _parse_resources, and changing the text slightly:

```Python
logger.debug("Configuring CORS for '%s' with resources: %s", app.name, resources) 
```

michalbachowski replied:
+1

CoryDolphin commented on flask_cors/__init__.py:
```diff
@@ -298,6 +312,19 @@ def _reg_length(maybe_regex):
+    try:
+        return regexp.pattern
+    except AttributeError:
+        return str(regexp)
```
I love the idea of printing out user-compiled regexs as the original pattern for easier reading. Two quick ideas I would love your feedback on:

1.  I did a similar check using the RegexObject type which I sniff at runtime (since the type is not actually part of the Public API). I don't feel strongly as to whether we use the 'ask for forgiveness' pattern of catching the exception, or checking the instance type, but I think we should be consistent

    ```Python
         def _reg_length(maybe_regex):
             if isinstance(maybe_regex, RegexObject):
                 return len(maybe_regex.pattern)
             else:
                 return len(maybe_regex)
    ```

2. If we change this function to be called '_get_regexp_pattern', It can be shared in both places, which is wonderful. If you would be willing to name-change, and update the usage in parsing resources to be:

```Python
         # Sort by the String length, or regex pattern length if compiled
         pattern_length = lambda pair: len(_get_regexp_pattern(pair[0]))
         return sorted(resources,
                       key=pattern_length,
                       reverse=True)
```

Since this one is a bit more involved, I am happy to take on the work of refactoring these two, since I don't want it to hold you up!

michalbachowski replied:
I've changes `_stringify_regexp()` to `_get_regexp_pattern()`. However I'm in favour of more pythonic "ask for forgiveness" rather than checking for type.

CoryDolphin replied:
Sounds great to me!

CoryDolphin commented:
@michalbachowski This is seriously awesome work! Thank you for your detailed work adding so much logging and digging through all of the code! I added a bunch of comments and a few questions which I would appreciate if you could take a few minutes to look at. If you don't have time/want to address all of the comments, just let me know and I will cherry-pick your commit and make the changes myself. 

At a high level, my view is that logging in Flask-Cors should do two things:
1. At the INFO level, make it easy to see how Flask-Cors is changing the responses your application is sending
2. At the DEBUG level, make it easy to see how the options you have passed to Flask-CORS resulted in these headers being sent. 

Do you think those make sense as goals? I am not as familiar with Flask's logging philosophy, and cannot seem to find too much information about it. 

As you said, I think we should convert all of the calls to `.format` to use logging's built in String formatter `%` strings, as it will not only fix Py26, but also remove some boilerplate calls to .format.

Thanks again for the awesome contribution! You rock!

michalbachowski commented:
I've implemented almost all what you suggested (except switching from 'ask for forgiveness' to type checking) and added default NullHandler() to logger as well.

In order to make this new logging easy to use, new section to README should be added, describing how to turn it on. I think the [Flask Error Handling](http://flask.pocoo.org/docs/0.10/errorhandling/#other-libraries) documention might be a good start.

CoryDolphin commented:
Awesome! I'm on my  way to work right now and will take a deeper look
tonight!

On Tue, Jan 20, 2015 at 2:46 AM, Michał Bachowski <notifications@github.com>
wrote:

> I've implemented almost all what you suggested (except switching from 'ask
> for forgiveness' to type checking) and added default NullHandler() to
> logger as well.
>
> In order to make this new logging easy to use, new section to README
> should be added, describing how to turn it on. I think the Flask Error
> Handling <http://flask.pocoo.org/docs/0.10/errorhandling/#other-libraries>
> documention might be a good start.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CoryDolphin/flask-cors/pull/99#issuecomment-70616023>.
>

CoryDolphin commented:
This looks great! Sorry for the delay!

Merging and will bump versions, and a few other little refactors to make consistent use of _get_regexp_pattern.

michalbachowski commented:
Great - looking forward to new version :-)
